pkg/blobstore/spanner_blob_access.go

```diff
@@ -0,0 +1,975 @@
+			}
+			if err := ba.delete(ctx, key); err == nil {
+				spannerMalformedBlobDeletedCount.WithLabelValues(beType).Inc()
+				log.Printf("Blob %s was malformed and has been deleted from Spanner/GCS successfully", digest.String())
+			} else {
+				spannerMalformedBlobDeleteFailedCount.WithLabelValues(beType).Inc()
+				log.Printf("Blob %s was malformed and could not be deleted from Spanner/GCS: %v", digest.String(), err)
+			}
+		}
+	}
+
+	var b buffer.Buffer
+	if sz > maxSize {
+		// We gotta go get it from GCS
+		obj := ba.gcsBucket.Object(key)
+
+		r, err := obj.NewReader(ctx)
+		if err != nil {
+			// If we couldn't read the bucket, then let's delete it from spanner (and from gcs if we can!)
+			if err2 := ba.delete(ctx, key); err2 == nil {
+				gcsFailedReadDeletedBlobCount.Inc()
+				log.Printf("Blob %s was inaccessible in GCS (due to %v) and has been deleted from Spanner/GCS successfully", digest.String(), err)
+			} else {
+				gcsFailedReadDeleteBlobFailedCount.Inc()
+				log.Printf("Blob %s was inaccessible in GCS (due to %v) and could not be deleted from Spanner/GCS: %s", digest.String(), err, err2)
+			}
+			return buffer.NewBufferFromError(err)
+		}
+		b = ba.readBufferFactory.NewBufferFromReader(digest, r, validationFunc)
+		b = buffer.WithErrorHandler(
+			b,
+			&spannerErrorHandler{
+				start:  time.Now(),
+				sType:  BEMap[ba.storageType],
+				beType: BE_GCS,
+				op:     BE_GET,
+			})
+	} else {
+		// It's inline, so return it directly
+		b = ba.readBufferFactory.NewBufferFromByteSlice(digest, s.InlineData, validationFunc)
+	}
+	_, err = b.GetSizeBytes()
+	if err != nil {
+		log.Printf("GET ERROR: key %s, type %v: %v", key, ba.storageType, err)
+	}
+	if ba.refChan != nil && now.After(s.ReferenceTime.Add(ba.refUpdateThresh)) {
+		log.Printf("GET: scheduling touch reftime for key %s reftime %s", key, s.ReferenceTime)
+		ba.refChan <- key
+	}
+	return b
+}
+
+func (ba *spannerBlobAccess) Put(ctx context.Context, digest digest.Digest, b buffer.Buffer) error {
+	//log.Printf("SpannerBlobAccess type %#+v PUT digest %s", ba.storageType, digest)
+	if err := util.StatusFromContext(ctx); err != nil {
+		b.Discard()
+		return err
+	}
+
+	// If we're bigger than 10MB, we have to offload to GCS
+	size, err := b.GetSizeBytes()
+	if err != nil {
+		log.Printf("Put Blob %s: can't get size: %v", digest, err)
+		b.Discard()
+		return err
+	}
+
+	key := ba.digestToKey(digest)
+	//log.Printf("SpannerBlobAccess PUT key is %s, size %d", key, size)
+
+	var inlineData []byte = nil
+	if ba.storageType == pb.StorageType_CASTORE {
+		statedSize, err := ba.keyToBlobSize(key)
+		if err != nil {
+			spannerMalformedKeyCount.Inc()
+			log.Printf("Put Blob: invalid digest size, key %s: %v", key, err)
+			b.Discard()
+			return err
+		}
+		if statedSize != size {
+			log.Printf("Put Blob: MISMATCH between buffer size (%d) and digest size (%d)", size, statedSize)
+			b.Discard()
+			return fmt.Errorf("Invalid contents digest %s size %d", digest, size)
+		}
```

Comment:
Do we really need code like this? `pkg/blobstore/buffer` already has logic for validating data integrity.

Reply:
Removed.